miner/Miner.hs

```diff
@@ -244,29 +226,60 @@ scheme env = case cmd $ args env of
     CPU e -> cpu (version $ args env) e (gen env)
     GPU _ -> gpu
 
--- | Submit a new `BlockHeader` to mine (i.e. to determine a valid `Nonce`).
+-- | Attempt to get new work while obeying a sane retry policy.
 --
-submit :: Env -> WorkBytes -> IO ()
-submit (work -> w) bs = atomically $
-    isEmptyTMVar w >>= bool (void $ swapTMVar w bs) (putTMVar w bs)
+getWork :: Env -> IO (Maybe WorkBytes)
+getWork e = retrying policy (\_ -> pure . isNothing) $ const f
+  where
+    -- | If we wait longer than the average block time and still can't get
+    -- anything, then there's no point in continuing to wait.
+    --
+    policy :: RetryPolicy
+    policy = exponentialBackoff 500000 <> limitRetries 7
+
+    f :: IO (Maybe WorkBytes)
+    f = hush <$> runClientM (workClient v (chainid a) $ miner a) (ClientEnv m u Nothing)
+
+    a = args e
+    v = version a
+    m = mgr e
+    u = _url $ coordinator a
 
--- | A supervisor thread that listens for new work and supervises mining threads.
+-- | A supervisor thread that listens for new work and manages mining threads.
 --
-mining :: (TargetBytes -> HeaderBytes -> IO HeaderBytes) -> Env -> IO ()
-mining go e = do
-    bs <- atomically . takeTMVar $ work e
-    race newWork (suncurry go $ unWorkBytes bs) >>= traverse_ miningSuccess
-    mining go e
+mining :: (TargetBytes -> HeaderBytes -> IO HeaderBytes) -> Env -> WorkBytes -> IO ()
+mining go e wb = do
+    race updateSignal (go tbytes hbytes) >>= traverse_ miningSuccess
+    getWork e >>= traverse_ (mining go e)
   where
-    -- | Wait for new work to come in from a `submit` call. `readTMVar` has the
-    -- effect of "waiting patiently" and will not cook the CPU.
-    --
-    newWork :: IO ()
-    newWork = void . atomically . readTMVar $ work e
+    T3 cbytes tbytes hbytes = unWorkBytes wb
+
+    a :: ClientArgs
+    a = args e
+
+    -- TODO Rework to use Servant's streaming? Otherwise I can't use the
+    -- convenient client function here.
+    updateSignal :: IO ()
+    updateSignal = withEvents req (mgr e) (void . SP.head_ . SP.filter realEvent)
```

Comment:
The first Server-Sent-Event to appear here in the `Stream` will kill the race, thus prompting a new `getWork` call.